File: python/compress_zip/file_format.py
"""
CZIPv1 file format for neural compression.

Outer envelope structure:
- 6 bytes: magic "CZIPv1"
- 1 byte: flags (bit0=is_multifile, bit1=training_marker, bit2=reserved)
- 1 byte: codec_id (0=zstd, 1=brotli)
- 4 bytes: uncompressed_len_bytes (u32 LE) - original text byte length
- 4 bytes: header_len_bytes (u32 LE) - inner header length before compression
- payload: compressed bytes of [inner_header || chunk_data]

Inner header structure:
- u32: model_id_hash
- u32: chunk_count
- u16: last_chunk_tokens (1..64)
- u8: tokenizer_id
- u8: reserved
- u16[chunk_count]: chunk_byte_len array
- [chunk_data]: concatenated chunk payloads
"""

import struct
import zlib
from dataclasses import dataclass, field
from typing import List, Tuple, Optional
from pathlib import Path
from enum import IntEnum

try:
    import zstandard as zstd
    HAS_ZSTD = True
except ImportError:
    HAS_ZSTD = False

try:
    import brotli
    HAS_BROTLI = True
except ImportError:
    HAS_BROTLI = False


# Magic number for file format identification
CZIP_MAGIC = b'CZIPv1'  # 6 bytes
OUTER_HEADER_SIZE = 16  # Fixed outer header size

# Chunk configuration
TOKENS_PER_CHUNK = 64  # All chunks except last have exactly 64 tokens


class Codec(IntEnum):
    """Compression codec identifiers."""
    ZSTD = 0
    BROTLI = 1


class Flags(IntEnum):
    """Flag bit positions."""
    IS_MULTIFILE = 0
    TRAINING_MARKER = 1
    RESERVED = 2


@dataclass
class CZIPv1OuterHeader:
    """
    CZIPv1 outer envelope header.

    Layout (16 bytes):
    - magic: 6 bytes ("CZIPv1")
    - flags: 1 byte (bitfield)
    - codec_id: 1 byte (0=zstd, 1=brotli)
    - uncompressed_len: 4 bytes (u32 LE, original text byte length)
    - header_len: 4 bytes (u32 LE, inner header length before compression)
    """
    magic: bytes = CZIP_MAGIC
    flags: int = 0
    codec_id: int = Codec.ZSTD
    uncompressed_len: int = 0  # Original text byte length
    header_len: int = 0  # Inner header length before compression

    def is_multifile(self) -> bool:
        return bool(self.flags & (1 << Flags.IS_MULTIFILE))

    def set_multifile(self, value: bool):
        if value:
            self.flags |= (1 << Flags.IS_MULTIFILE)
        else:
            self.flags &= ~(1 << Flags.IS_MULTIFILE)

    def to_bytes(self) -> bytes:
        """Serialize outer header to bytes."""
        return struct.pack(
            '<6sBBII',
            self.magic,
            self.flags,
            self.codec_id,
            self.uncompressed_len,
            self.header_len,
        )

    @classmethod
    def from_bytes(cls, data: bytes) -> "CZIPv1OuterHeader":
        """Deserialize outer header from bytes."""
        if len(data) < OUTER_HEADER_SIZE:
            raise ValueError(f"Outer header too short: {len(data)} < {OUTER_HEADER_SIZE}")

        magic, flags, codec_id, uncompressed_len, header_len = struct.unpack(
            '<6sBBII',
            data[:OUTER_HEADER_SIZE]
        )

        if magic != CZIP_MAGIC:
            raise ValueError(f"Invalid magic: {magic!r}, expected {CZIP_MAGIC!r}")

        return cls(
            magic=magic,
            flags=flags,
            codec_id=codec_id,
            uncompressed_len=uncompressed_len,
            header_len=header_len,
        )


@dataclass
class CZIPv1InnerHeader:
    """
    CZIPv1 inner header (before outer compression).

    Layout:
    - model_id_hash: 4 bytes (u32 LE, for mismatch detection, 0 if unused)
    - chunk_count: 4 bytes (u32 LE)
    - last_chunk_tokens: 2 bytes (u16 LE, 1..64)
    - tokenizer_id: 1 byte (u8, 0 for default)
    - reserved: 1 byte (u8)
    - chunk_byte_lens: chunk_count * 2 bytes (u16 LE each)
    """
    model_id_hash: int = 0
    chunk_count: int = 0
    last_chunk_tokens: int = TOKENS_PER_CHUNK
    tokenizer_id: int = 0
    reserved: int = 0
    chunk_byte_lens: List[int] = field(default_factory=list)

    def header_size(self) -> int:
        """Get inner header size in bytes (excluding chunk data)."""
        return 12 + 2 * self.chunk_count

    def to_bytes(self) -> bytes:
        """Serialize inner header to bytes."""
        buf = struct.pack(
            '<IIHBB',
            self.model_id_hash,
            self.chunk_count,
            self.last_chunk_tokens,
            self.tokenizer_id,
            self.reserved,
        )
        # Append chunk byte lengths
        for byte_len in self.chunk_byte_lens:
            buf += struct.pack('<H', byte_len)
        return buf

    @classmethod
    def from_bytes(cls, data: bytes) -> "CZIPv1InnerHeader":
        """Deserialize inner header from bytes."""
        if len(data) < 12:
            raise ValueError(f"Inner header too short: {len(data)} < 12")

        model_id_hash, chunk_count, last_chunk_tokens, tokenizer_id, reserved = struct.unpack(
            '<IIHBB',
            data[:12]
        )

        expected_len = 12 + 2 * chunk_count
        if len(data) < expected_len:
            raise ValueError(f"Inner header too short for {chunk_count} chunks")

        chunk_byte_lens = []
        offset = 12
        for _ in range(chunk_count):
            byte_len = struct.unpack('<H', data[offset:offset + 2])[0]
            chunk_byte_lens.append(byte_len)
            offset += 2

        return cls(
            model_id_hash=model_id_hash,
            chunk_count=chunk_count,
            last_chunk_tokens=last_chunk_tokens,
            tokenizer_id=tokenizer_id,
            reserved=reserved,
            chunk_byte_lens=chunk_byte_lens,
        )


def compress_payload(data: bytes, codec: Codec, level: Optional[int] = None) -> bytes:
    """Compress payload using specified codec."""
    if codec == Codec.ZSTD:
        if not HAS_ZSTD:
            raise RuntimeError("zstandard library not installed")
        lvl = level if level is not None else 22
        cctx = zstd.ZstdCompressor(level=lvl)
        return cctx.compress(data)
    elif codec == Codec.BROTLI:
        if not HAS_BROTLI:
            raise RuntimeError("brotli library not installed")
        lvl = level if level is not None else 11
        return brotli.compress(data, quality=lvl)
    else:
        raise ValueError(f"Unknown codec: {codec}")


def decompress_payload(data: bytes, codec: Codec) -> bytes:
    """Decompress payload using specified codec."""
    if codec == Codec.ZSTD:
        if not HAS_ZSTD:
            raise RuntimeError("zstandard library not installed")
        dctx = zstd.ZstdDecompressor()
        return dctx.decompress(data)
    elif codec == Codec.BROTLI:
        if not HAS_BROTLI:
            raise RuntimeError("brotli library not installed")
        return brotli.decompress(data)
    else:
        raise ValueError(f"Unknown codec: {codec}")


class CompressedFile:
    """
    Read/write CZIPv1 compressed files.
    """

    def __init__(self):
        self.outer_header: CZIPv1OuterHeader = CZIPv1OuterHeader()
        self.inner_header: CZIPv1InnerHeader = CZIPv1InnerHeader()
        self.chunks: List[bytes] = []  # Raw arithmetic-coded chunk data

    @classmethod
    def create(
        cls,
        model_id_hash: int = 0,
        tokenizer_id: int = 0,
        codec: Codec = Codec.ZSTD,
        is_multifile: bool = False,
    ) -> "CompressedFile":
        """Create a new compressed file."""
        cf = cls()
        cf.outer_header.codec_id = codec
        cf.outer_header.set_multifile(is_multifile)
        cf.inner_header.model_id_hash = model_id_hash
        cf.inner_header.tokenizer_id = tokenizer_id
        return cf

    def add_chunk(self, compressed_data: bytes, is_last: bool = False, last_chunk_tokens: int = TOKENS_PER_CHUNK):
        """
        Add a compressed chunk.

        Args:
            compressed_data: Arithmetic-coded bytes for this chunk
            is_last: Whether this is the last chunk
            last_chunk_tokens: Number of tokens in last chunk (1..64)
        """
        if len(compressed_data) > 65535:
            raise ValueError(f"Chunk too large: {len(compressed_data)} > 65535 bytes")

        self.chunks.append(compressed_data)
        self.inner_header.chunk_byte_lens.append(len(compressed_data))
        self.inner_header.chunk_count = len(self.chunks)

        if is_last:
            self.inner_header.last_chunk_tokens = last_chunk_tokens

    def finalize(self, uncompressed_text_len: int, compression_level: Optional[int] = None):
        """
        Finalize the file for writing.

        Args:
            uncompressed_text_len: Original text length in bytes (before tokenization)
            compression_level: Optional compression level override
        """
        self.outer_header.uncompressed_len = uncompressed_text_len
        self.outer_header.header_len = self.inner_header.header_size()

    def _build_payload(self) -> bytes:
        """Build the uncompressed payload (inner_header || chunk_data)."""
        payload = self.inner_header.to_bytes()
        for chunk in self.chunks:
            payload += chunk
        return payload

    def write(self, path: Path | str, compression_level: Optional[int] = None):
        """Write compressed file to disk."""
        path = Path(path)

        # Build and compress payload
        payload = self._build_payload()
        compressed_payload = compress_payload(
            payload,
            Codec(self.outer_header.codec_id),
            compression_level
        )

        # Update header with final sizes
        self.outer_header.header_len = self.inner_header.header_size()

        with open(path, 'wb') as f:
            f.write(self.outer_header.to_bytes())
            f.write(compressed_payload)

    def to_bytes(self, compression_level: Optional[int] = None) -> bytes:
        """Serialize to bytes."""
        payload = self._build_payload()
        compressed_payload = compress_payload(
            payload,
            Codec(self.outer_header.codec_id),
            compression_level
        )
        return self.outer_header.to_bytes() + compressed_payload

    @classmethod
    def read(cls, path: Path | str) -> "CompressedFile":
        """Read compressed file from disk."""
        path = Path(path)

        with open(path, 'rb') as f:
            data = f.read()

        return cls.from_bytes(data)

    @classmethod
    def from_bytes(cls, data: bytes) -> "CompressedFile":
        """Deserialize from bytes."""
        cf = cls()

        # Parse outer header
        cf.outer_header = CZIPv1OuterHeader.from_bytes(data[:OUTER_HEADER_SIZE])

        # Decompress payload
        compressed_payload = data[OUTER_HEADER_SIZE:]
        payload = decompress_payload(compressed_payload, Codec(cf.outer_header.codec_id))

        # Parse inner header
        cf.inner_header = CZIPv1InnerHeader.from_bytes(payload)

        # Extract chunks
        offset = cf.inner_header.header_size()
        cf.chunks = []
        for byte_len in cf.inner_header.chunk_byte_lens:
            chunk_data = payload[offset:offset + byte_len]
            if len(chunk_data) != byte_len:
                raise ValueError(f"Chunk data truncated: {len(chunk_data)} < {byte_len}")
            cf.chunks.append(chunk_data)
            offset += byte_len

        return cf

    def get_total_tokens(self) -> int:
        """Get total number of tokens across all chunks."""
        if self.inner_header.chunk_count == 0:
            return 0
        # All chunks except last have TOKENS_PER_CHUNK tokens
        full_chunks = self.inner_header.chunk_count - 1
        return full_chunks * TOKENS_PER_CHUNK + self.inner_header.last_chunk_tokens

    def get_chunk_tokens(self, chunk_idx: int) -> int:
        """Get number of tokens in a specific chunk."""
        if chunk_idx < 0 or chunk_idx >= self.inner_header.chunk_count:
            raise IndexError(f"Chunk index {chunk_idx} out of range")
        if chunk_idx == self.inner_header.chunk_count - 1:
            return self.inner_header.last_chunk_tokens
        return TOKENS_PER_CHUNK

    def get_compressed_size(self) -> int:
        """Get total compressed size in bytes (outer header + compressed payload)."""
        payload = self._build_payload()
        compressed = compress_payload(payload, Codec(self.outer_header.codec_id))
        return OUTER_HEADER_SIZE + len(compressed)


# Test cases
if __name__ == "__main__":
    import tempfile

    print("Testing CZIPv1 file format...")

    # Test outer header serialization
    outer = CZIPv1OuterHeader(
        codec_id=Codec.ZSTD,
        uncompressed_len=1000,
        header_len=20,
    )
    outer.set_multifile(True)
    outer_bytes = outer.to_bytes()
    print(f"Outer header size: {len(outer_bytes)} bytes")
    assert len(outer_bytes) == OUTER_HEADER_SIZE

    outer2 = CZIPv1OuterHeader.from_bytes(outer_bytes)
    assert outer2.codec_id == outer.codec_id
    assert outer2.uncompressed_len == outer.uncompressed_len
    assert outer2.is_multifile() == True
    print("Outer header roundtrip passed!")

    # Test inner header serialization
    inner = CZIPv1InnerHeader(
        model_id_hash=0x12345678,
        chunk_count=3,
        last_chunk_tokens=32,
        tokenizer_id=0,
        chunk_byte_lens=[100, 150, 80],
    )
    inner_bytes = inner.to_bytes()
    print(f"Inner header size: {len(inner_bytes)} bytes (expected {inner.header_size()})")
    assert len(inner_bytes) == inner.header_size()

    inner2 = CZIPv1InnerHeader.from_bytes(inner_bytes)
    assert inner2.model_id_hash == inner.model_id_hash
    assert inner2.chunk_count == inner.chunk_count
    assert inner2.last_chunk_tokens == inner.last_chunk_tokens
    assert inner2.chunk_byte_lens == inner.chunk_byte_lens
    print("Inner header roundtrip passed!")

    # Test file read/write (only if compression libs available)
    if HAS_ZSTD:
        with tempfile.TemporaryDirectory() as tmpdir:
            test_path = Path(tmpdir) / "test.czip"

            # Create file
            cf = CompressedFile.create(
                model_id_hash=0xDEADBEEF,
                tokenizer_id=0,
                codec=Codec.ZSTD,
            )

            # Add some chunks
            cf.add_chunk(b'\x00' * 100)
            cf.add_chunk(b'\x11' * 150)
            cf.add_chunk(b'\x22' * 80, is_last=True, last_chunk_tokens=32)

            cf.finalize(uncompressed_text_len=1000)

            # Write
            cf.write(test_path)
            print(f"Wrote file to {test_path}")

            # Read back
            cf2 = CompressedFile.read(test_path)

            assert cf2.inner_header.model_id_hash == cf.inner_header.model_id_hash
            assert cf2.inner_header.chunk_count == cf.inner_header.chunk_count
            assert cf2.inner_header.last_chunk_tokens == cf.inner_header.last_chunk_tokens
            assert len(cf2.chunks) == len(cf.chunks)

            for i, (d1, d2) in enumerate(zip(cf.chunks, cf2.chunks)):
                assert d1 == d2, f"Chunk {i} data mismatch"

            print("File roundtrip passed!")

            # Test total tokens
            assert cf2.get_total_tokens() == 64 + 64 + 32
            print(f"Total tokens: {cf2.get_total_tokens()}")

        print("\nAll CZIPv1 file format tests passed!")
    else:
        print("\nSkipping file tests (zstandard not installed)")
        print("Install with: pip install zstandard")

File: rust/src/file_format.rs
//! CZIPv1 file format for compressed data.
//!
//! Outer envelope (16 bytes + compressed payload):
//! - magic: [u8; 6] = b"CZIPv1"
//! - flags: u8 (bit0=is_multifile, bit1=training_marker, bit2=reserved)
//! - codec_id: u8 (0=zstd, 1=brotli)
//! - uncompressed_len: u32 LE (original text byte length)
//! - header_len: u32 LE (inner header length before compression)
//! - payload: compressed bytes of [inner_header || chunk_data]
//!
//! Inner header (before compression):
//! - model_id_hash: u32 LE
//! - chunk_count: u32 LE
//! - last_chunk_tokens: u16 LE (1..64)
//! - tokenizer_id: u8
//! - reserved: u8
//! - chunk_byte_lens: [u16; chunk_count] LE

use std::io::{self, Read, Write};

/// File format magic bytes
pub const MAGIC: &[u8; 6] = b"CZIPv1";

/// Outer header size in bytes
pub const OUTER_HEADER_SIZE: usize = 16;

/// Fixed tokens per chunk (except last chunk)
pub const TOKENS_PER_CHUNK: u16 = 64;

/// Compression codec identifiers
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum Codec {
    Zstd = 0,
    Brotli = 1,
}

impl TryFrom<u8> for Codec {
    type Error = io::Error;

    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            0 => Ok(Codec::Zstd),
            1 => Ok(Codec::Brotli),
            _ => Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!("Unknown codec: {}", value),
            )),
        }
    }
}

/// Flag bit positions
pub mod flags {
    pub const IS_MULTIFILE: u8 = 0;
    pub const TRAINING_MARKER: u8 = 1;
    pub const RESERVED: u8 = 2;
}

/// CZIPv1 outer envelope header.
#[derive(Debug, Clone)]
pub struct CZIPv1OuterHeader {
    pub magic: [u8; 6],
    pub flags: u8,
    pub codec_id: Codec,
    pub uncompressed_len: u32,  // Original text byte length
    pub header_len: u32,        // Inner header length before compression
}

impl CZIPv1OuterHeader {
    /// Create a new outer header.
    pub fn new(codec: Codec, uncompressed_len: u32) -> Self {
        Self {
            magic: *MAGIC,
            flags: 0,
            codec_id: codec,
            uncompressed_len,
            header_len: 0,
        }
    }

    /// Check if multifile flag is set.
    pub fn is_multifile(&self) -> bool {
        (self.flags & (1 << flags::IS_MULTIFILE)) != 0
    }

    /// Set the multifile flag.
    pub fn set_multifile(&mut self, value: bool) {
        if value {
            self.flags |= 1 << flags::IS_MULTIFILE;
        } else {
            self.flags &= !(1 << flags::IS_MULTIFILE);
        }
    }

    /// Read header from bytes.
    pub fn from_bytes(data: &[u8]) -> io::Result<Self> {
        if data.len() < OUTER_HEADER_SIZE {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "Outer header too short",
            ));
        }

        let mut magic = [0u8; 6];
        magic.copy_from_slice(&data[0..6]);

        if &magic != MAGIC {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!("Invalid magic bytes: {:?}", magic),
            ));
        }

        let flags = data[6];
        let codec_id = Codec::try_from(data[7])?;
        let uncompressed_len = u32::from_le_bytes(data[8..12].try_into().unwrap());
        let header_len = u32::from_le_bytes(data[12..16].try_into().unwrap());

        Ok(Self {
            magic,
            flags,
            codec_id,
            uncompressed_len,
            header_len,
        })
    }

    /// Write header to bytes.
    pub fn to_bytes(&self) -> Vec<u8> {
        let mut buf = Vec::with_capacity(OUTER_HEADER_SIZE);
        buf.extend_from_slice(&self.magic);
        buf.push(self.flags);
        buf.push(self.codec_id as u8);
        buf.extend_from_slice(&self.uncompressed_len.to_le_bytes());
        buf.extend_from_slice(&self.header_len.to_le_bytes());
        buf
    }
}

/// CZIPv1 inner header (before outer compression).
#[derive(Debug, Clone)]
pub struct CZIPv1InnerHeader {
    pub model_id_hash: u32,
    pub chunk_count: u32,
    pub last_chunk_tokens: u16,
    pub tokenizer_id: u8,
    pub reserved: u8,
    pub chunk_byte_lens: Vec<u16>,
}

impl CZIPv1InnerHeader {
    /// Create a new inner header.
    pub fn new(model_id_hash: u32, tokenizer_id: u8) -> Self {
        Self {
            model_id_hash,
            chunk_count: 0,
            last_chunk_tokens: TOKENS_PER_CHUNK,
            tokenizer_id,
            reserved: 0,
            chunk_byte_lens: Vec::new(),
        }
    }

    /// Get inner header size in bytes (excluding chunk data).
    pub fn header_size(&self) -> usize {
        12 + 2 * self.chunk_count as usize
    }

    /// Read header from bytes.
    pub fn from_bytes(data: &[u8]) -> io::Result<Self> {
        if data.len() < 12 {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "Inner header too short",
            ));
        }

        let model_id_hash = u32::from_le_bytes(data[0..4].try_into().unwrap());
        let chunk_count = u32::from_le_bytes(data[4..8].try_into().unwrap());
        let last_chunk_tokens = u16::from_le_bytes(data[8..10].try_into().unwrap());
        let tokenizer_id = data[10];
        let reserved = data[11];

        let expected_len = 12 + 2 * chunk_count as usize;
        if data.len() < expected_len {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!("Inner header too short for {} chunks", chunk_count),
            ));
        }

        let mut chunk_byte_lens = Vec::with_capacity(chunk_count as usize);
        let mut offset = 12;
        for _ in 0..chunk_count {
            let byte_len = u16::from_le_bytes(data[offset..offset + 2].try_into().unwrap());
            chunk_byte_lens.push(byte_len);
            offset += 2;
        }

        Ok(Self {
            model_id_hash,
            chunk_count,
            last_chunk_tokens,
            tokenizer_id,
            reserved,
            chunk_byte_lens,
        })
    }

    /// Write header to bytes.
    pub fn to_bytes(&self) -> Vec<u8> {
        let mut buf = Vec::with_capacity(self.header_size());
        buf.extend_from_slice(&self.model_id_hash.to_le_bytes());
        buf.extend_from_slice(&self.chunk_count.to_le_bytes());
        buf.extend_from_slice(&self.last_chunk_tokens.to_le_bytes());
        buf.push(self.tokenizer_id);
        buf.push(self.reserved);
        for &byte_len in &self.chunk_byte_lens {
            buf.extend_from_slice(&byte_len.to_le_bytes());
        }
        buf
    }
}

/// A single encoded chunk.
#[derive(Debug, Clone)]
pub struct EncodedChunk {
    pub data: Vec<u8>,
}

/// Complete compressed file (uncompressed representation).
#[derive(Debug, Clone)]
pub struct CompressedFile {
    pub outer_header: CZIPv1OuterHeader,
    pub inner_header: CZIPv1InnerHeader,
    pub chunks: Vec<EncodedChunk>,
}

impl CompressedFile {
    /// Create a new compressed file.
    pub fn new(
        model_id_hash: u32,
        tokenizer_id: u8,
        codec: Codec,
        uncompressed_text_len: u32,
    ) -> Self {
        Self {
            outer_header: CZIPv1OuterHeader::new(codec, uncompressed_text_len),
            inner_header: CZIPv1InnerHeader::new(model_id_hash, tokenizer_id),
            chunks: Vec::new(),
        }
    }

    /// Add a chunk.
    pub fn add_chunk(&mut self, data: Vec<u8>, is_last: bool, last_chunk_tokens: u16) {
        assert!(data.len() <= u16::MAX as usize, "Chunk too large");

        self.inner_header.chunk_byte_lens.push(data.len() as u16);
        self.inner_header.chunk_count = self.inner_header.chunk_byte_lens.len() as u32;

        if is_last {
            self.inner_header.last_chunk_tokens = last_chunk_tokens;
        }

        self.chunks.push(EncodedChunk { data });
    }

    /// Build the uncompressed payload (inner_header || chunk_data).
    fn build_payload(&self) -> Vec<u8> {
        let mut payload = self.inner_header.to_bytes();
        for chunk in &self.chunks {
            payload.extend_from_slice(&chunk.data);
        }
        payload
    }

    /// Get total tokens across all chunks.
    pub fn get_total_tokens(&self) -> u32 {
        if self.inner_header.chunk_count == 0 {
            return 0;
        }
        let full_chunks = self.inner_header.chunk_count - 1;
        full_chunks * TOKENS_PER_CHUNK as u32 + self.inner_header.last_chunk_tokens as u32
    }

    /// Get token count for a specific chunk.
    pub fn get_chunk_tokens(&self, chunk_idx: usize) -> u16 {
        if chunk_idx == self.inner_header.chunk_count as usize - 1 {
            self.inner_header.last_chunk_tokens
        } else {
            TOKENS_PER_CHUNK
        }
    }

    /// Read from bytes (requires decompression).
    ///
    /// Note: This expects already-decompressed payload bytes after the outer header.
    /// For full file reading with compression, use `read_from_compressed`.
    pub fn read_from_decompressed(outer_header: CZIPv1OuterHeader, payload: &[u8]) -> io::Result<Self> {
        let inner_header = CZIPv1InnerHeader::from_bytes(payload)?;

        let mut chunks = Vec::with_capacity(inner_header.chunk_count as usize);
        let mut offset = inner_header.header_size();

        for &byte_len in &inner_header.chunk_byte_lens {
            let end = offset + byte_len as usize;
            if payload.len() < end {
                return Err(io::Error::new(
                    io::ErrorKind::InvalidData,
                    "Chunk data truncated",
                ));
            }
            chunks.push(EncodedChunk {
                data: payload[offset..end].to_vec(),
            });
            offset = end;
        }

        Ok(Self {
            outer_header,
            inner_header,
            chunks,
        })
    }

    /// Write to bytes (uncompressed payload, for external compression).
    ///
    /// Returns (outer_header_bytes, payload_bytes) for external compression.
    pub fn to_bytes_uncompressed(&self) -> (Vec<u8>, Vec<u8>) {
        let mut outer = self.outer_header.clone();
        outer.header_len = self.inner_header.header_size() as u32;

        (outer.to_bytes(), self.build_payload())
    }

    /// Write to a writer (without compression - for testing).
    ///
    /// Note: In production, the payload should be compressed with zstd or brotli.
    pub fn write_to_uncompressed<W: Write>(&self, writer: &mut W) -> io::Result<()> {
        let (outer_bytes, payload) = self.to_bytes_uncompressed();
        writer.write_all(&outer_bytes)?;
        writer.write_all(&payload)?;
        Ok(())
    }

    /// Read from a reader (without decompression - for testing).
    ///
    /// Note: In production, the payload should be decompressed first.
    pub fn read_from_uncompressed<R: Read>(reader: &mut R) -> io::Result<Self> {
        let mut outer_buf = [0u8; OUTER_HEADER_SIZE];
        reader.read_exact(&mut outer_buf)?;
        let outer_header = CZIPv1OuterHeader::from_bytes(&outer_buf)?;

        let mut payload = Vec::new();
        reader.read_to_end(&mut payload)?;

        Self::read_from_decompressed(outer_header, &payload)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Cursor;

    #[test]
    fn test_outer_header_roundtrip() {
        let mut header = CZIPv1OuterHeader::new(Codec::Zstd, 1000);
        header.set_multifile(true);
        header.header_len = 20;

        let bytes = header.to_bytes();
        assert_eq!(bytes.len(), OUTER_HEADER_SIZE);

        let parsed = CZIPv1OuterHeader::from_bytes(&bytes).unwrap();
        assert_eq!(parsed.magic, *MAGIC);
        assert_eq!(parsed.codec_id, Codec::Zstd);
        assert_eq!(parsed.uncompressed_len, 1000);
        assert_eq!(parsed.header_len, 20);
        assert!(parsed.is_multifile());
    }

    #[test]
    fn test_inner_header_roundtrip() {
        let mut header = CZIPv1InnerHeader::new(0x12345678, 0);
        header.chunk_count = 3;
        header.last_chunk_tokens = 32;
        header.chunk_byte_lens = vec![100, 150, 80];

        let bytes = header.to_bytes();
        assert_eq!(bytes.len(), header.header_size());

        let parsed = CZIPv1InnerHeader::from_bytes(&bytes).unwrap();
        assert_eq!(parsed.model_id_hash, 0x12345678);
        assert_eq!(parsed.chunk_count, 3);
        assert_eq!(parsed.last_chunk_tokens, 32);
        assert_eq!(parsed.chunk_byte_lens, vec![100, 150, 80]);
    }

    #[test]
    fn test_header_roundtrip() {
        // Legacy test - using new format
        let mut file = CompressedFile::new(0xABCDEF00, 0, Codec::Zstd, 5000);
        file.add_chunk(vec![1, 2, 3], false, TOKENS_PER_CHUNK);
        file.add_chunk(vec![4, 5], true, 32);

        let (outer_bytes, payload) = file.to_bytes_uncompressed();

        let outer = CZIPv1OuterHeader::from_bytes(&outer_bytes).unwrap();
        let parsed = CompressedFile::read_from_decompressed(outer, &payload).unwrap();

        assert_eq!(parsed.inner_header.model_id_hash, 0xABCDEF00);
        assert_eq!(parsed.inner_header.chunk_count, 2);
        assert_eq!(parsed.inner_header.last_chunk_tokens, 32);
        assert_eq!(parsed.get_total_tokens(), 64 + 32);
    }

    #[test]
    fn test_file_roundtrip() {
        let mut file = CompressedFile::new(0x12345678, 1, Codec::Brotli, 2000);
        file.add_chunk(vec![1, 2, 3, 4, 5], false, TOKENS_PER_CHUNK);
        file.add_chunk(vec![6, 7, 8], false, TOKENS_PER_CHUNK);
        file.add_chunk(vec![9, 10], true, 48);

        let mut buf = Vec::new();
        file.write_to_uncompressed(&mut buf).unwrap();

        let mut cursor = Cursor::new(buf);
        let parsed = CompressedFile::read_from_uncompressed(&mut cursor).unwrap();

        assert_eq!(parsed.outer_header.codec_id, Codec::Brotli);
        assert_eq!(parsed.outer_header.uncompressed_len, 2000);
        assert_eq!(parsed.inner_header.model_id_hash, 0x12345678);
        assert_eq!(parsed.inner_header.tokenizer_id, 1);
        assert_eq!(parsed.inner_header.chunk_count, 3);
        assert_eq!(parsed.inner_header.last_chunk_tokens, 48);
        assert_eq!(parsed.chunks.len(), 3);
        assert_eq!(parsed.chunks[0].data, vec![1, 2, 3, 4, 5]);
        assert_eq!(parsed.chunks[1].data, vec![6, 7, 8]);
        assert_eq!(parsed.chunks[2].data, vec![9, 10]);
        assert_eq!(parsed.get_total_tokens(), 64 + 64 + 48);
        assert_eq!(parsed.get_chunk_tokens(0), 64);
        assert_eq!(parsed.get_chunk_tokens(1), 64);
        assert_eq!(parsed.get_chunk_tokens(2), 48);
    }
}
